crates/ruff_python_semantic/src/cfg/graph.rs

```diff
@@ -0,0 +1,282 @@
+        mut predicate: T,
+    ) -> impl Iterator<Item = BlockId> + 'a {
+        self.conditions()
+            .zip(self.targets())
+            .filter(move |(cond, _)| predicate(cond))
+            .map(|(_, block)| block)
+    }
+}
+
+/// Represents a condition to be tested in a multi-way branch
+#[derive(Debug, Clone)]
+pub enum Condition {
+    /// Unconditional edge
+    Always,
+}
+
+struct CFGBuilder<'stmt> {
+    /// Control flow graph under construction
+    cfg: CFG<'stmt>,
+    /// Current basic block index
+    current: BlockId,
+    /// Exit block index for current control flow
+    exit: BlockId,
+}
+
+impl<'stmt> CFGBuilder<'stmt> {
+    /// Returns [`CFGBuilder`] with vector of blocks initialized at given capacity and with both initial and terminal blocks populated.
+    fn with_capacity(capacity: usize) -> Self {
+        let mut blocks = IndexVec::with_capacity(capacity);
+        let initial = blocks.push(BlockData {
+            kind: BlockKind::Start,
+            ..BlockData::default()
+        });
+        let terminal = blocks.push(BlockData {
+            kind: BlockKind::Terminal,
+            ..BlockData::default()
+        });
+
+        Self {
+            cfg: CFG {
+                blocks,
+                initial,
+                terminal,
+            },
+            current: initial,
+            exit: terminal,
+        }
+    }
+
+    /// Runs the core logic for the builder.
+    fn process_stmts(&mut self, stmts: &'stmt [Stmt]) {
+        let start = 0;
+        for stmt in stmts {
+            let cache_exit = self.exit();
+            match stmt {
+                Stmt::FunctionDef(_)
+                | Stmt::ClassDef(_)
+                | Stmt::Assign(_)
+                | Stmt::AugAssign(_)
+                | Stmt::AnnAssign(_)
+                | Stmt::TypeAlias(_)
+                | Stmt::Import(_)
+                | Stmt::ImportFrom(_)
+                | Stmt::Global(_)
+                | Stmt::Nonlocal(_)
+                | Stmt::Expr(_)
+                | Stmt::Pass(_)
+                | Stmt::Delete(_)
+                | Stmt::IpyEscapeCommand(_) => {}
+                // Loops
+                Stmt::While(_) => {}
+                Stmt::For(_) => {}
+
+                // Switch statements
+                Stmt::If(_) => {}
+                Stmt::Match(_) => {}
+
+                // Exception handling statements
+                Stmt::Try(_) => {}
+                Stmt::With(_) => {}
+
+                // Jumps
+                Stmt::Return(_) => {}
+                Stmt::Break(_) => {}
+                Stmt::Continue(_) => {}
+                Stmt::Raise(_) => {}
+
+                // An `assert` is a mixture of a switch and a jump.
+                Stmt::Assert(_) => {}
+            }
+            // Restore exit
+            self.update_exit(cache_exit);
+        }
+        // Push remaining statements
+        if start < stmts.len() {
+            self.set_stmts(&stmts[start..]);
+        }
```

Comment:
This line isn't clear to me and the comment doesn't add much. What's this line about? Aren't we always setting all statements?